### ISSUE TITLE
UserWritten variable names should have priority over system-Inserted ones

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2424]
```agda
open import Common.Nat
open import Common.List

test : (As : List Set) (x : Nat) → Set
test As x = {!As!}  -- C-c C-c
```
Split on `As` produces
```agda
test [] x = {!!}
test (x ∷ As) x₁ = {!!}
```
but I'd rather expect
```agda
test [] x = {!!}
test (x₁ ∷ As) x = {!!}
```

**Comment by andreasabel** [comment]
Related: #819 (preserve underscores), #573 (preserve ellipsis), #1820 (shadowing ellipsis variables).

I think this can be fixed as follows:
1. Before splitting, we know which variables were `UserWritten` and their original names should be preserved in the pattern variables (check this!).  
2. The case splitting itself replaces some variables by patterns or dot patterns, we need to make sure that everything inserted by case splitting is labeled `Inserted`.
3. After the split, we do a clean-up walk over the clause telescope to rename conflicting identifiers.  All `UserWritten` names should be unique already (as coming from the linear original patterns).  The inserted name will then be renamed, avoiding the user-written ones.  
4. The printer for the clause will then be in a situation where it does not have to do any renaming anymore.

**Comment by andreasabel** [comment]
The above plan has not been fully implemented.  Actually, it is easier to build on the fix for #2018.  I will push a fix (also for #2729).
